Update and enliven advanced intro/review notebook (#594)
AlexsLemonade/training-modules: scRNA-seq-advanced/01-read_filter_normalize_scRNA.Rmd

# Path to mitochondrial genes table
mito_file <- file.path(ref_dir, "hs_mitochondrial_genes.tsv")
```

```{r outputs}
# Outputs ------------------------------------

# Directory and file to save output
normalized_dir <- file.path(data_dir, "normalized")

# create the directory if it does not exist 
if (!dir.exists(normalized_dir)) {
  dir.create(normalized_dir, recursive = TRUE)
}

# output RDS file for normalized data
output_sce_file <- file.path(normalized_dir, "gliobastoma_normalized_sce.rds")
```

## Reading Cell Ranger data

Whether the 10X Cell Ranger data is in Matrix Exchange format or in an HDF5 file, we can use the `read10xCounts()` function from the `DropletUtils` package to read the data into R and create a `SingleCellExperiment` object.

> (If you used something other than Cell Ranger to process the raw data, you would need to use a different function to read it in and create the `SingleCellExperiment` object.)

Reviewer: Might be worth linking this section - http://bioconductor.org/books/3.16/OSCA.intro/getting-scrna-seq-datasets.html#reading-counts-into-r

```{r read SCE, live=TRUE}
# read SCE from matrix directory
raw_sce <- DropletUtils::read10xCounts(raw_matrix_dir) 
```

Let's look at the contents of the object after reading it in:

```{r view SCE, live=TRUE}
# view SCE object
raw_sce
```

We can see from this summary that this `SingleCellExperiment` (SCE) object contains 36,601 rows, which correspond to the features (genes) that were analyzed, and 734,492 columns, which correspond to the possible barcode tags that were used in the experiment.
Note that not all of these barcode tags will have been used, and many of the features may never have been seen either.
One of our first steps will be to filter out barcodes that were never seen, or that may have only been seen in a droplet that did not contain a cell (an "empty droplet").

### Structure of the `SingleCellExperiment` object

> In addition to the main `counts` matrix, which you can see listed as an `assay` in the `SingleCellExperiment` object summary above, the SCE object can contain a number of other tables and matrices, each stored in a "slot" with a particular format. 

Reviewer: suggestions to make this a little less wordy, up to you

```suggestion
In addition to the main `counts` matrix, listed as an `assay` in the SCE summary above, the SCE object can contain a number of other tables and matrices, each stored in a "slot" with a particular format. 
```

The overall structure of the object can be seen in the figure below, which comes from an [OSCA Introduction chapter](http://bioconductor.org/books/3.16/OSCA.intro/the-singlecellexperiment-class.html).

![structure of a SingleCellExperiment object](diagrams/SingleCellExperiment.png)

We have just mentioned the main `assay` slot, which contains full matrices of data (such as transcript counts) with each row a feature and each column a cell. 

> There are also a couple of tables for metadata, and a slot to store reduced-dimension representations of the expression data.

Reviewer: ```suggestion
There are also a couple of tables for metadata, and a slot to store reduced-dimension representations (e.g. PCA and/or UMAP) of the expression data.
```

> We'll start with the `rowData` slot, which is a table of metadata for each feature in our object. 

Reviewer: Might also add "in this case our features are genes" for smoother transition into using "gene" terminology

Now we can look at how our SCE object has changed:

> ```{r zero-filtered SCE}

Reviewer: live?

raw_sce
```

But barcodes with zero counts are not the only ones that correspond to droplets without cells in them!

> Even if a droplet does not have a cell in it, there will often be spurious reads from RNA sequences that were present in the extracellular solution, whether from the original sample or cells that were disrupted during library preparation for single-cell analysis. 

Reviewer: small rephrases at your discretion - 

```suggestion
Even if a droplet does not have a cell in it, there will often be spurious reads from RNA sequences that were present in the extracellular solution, whether from the original sample or from cells that were disrupted during single-cell library preparation. 
```

This test was first proposed by [Lun _et al._ (2019)](https://doi.org/10.1186/s13059-019-1662-y) and implemented as `emptyDrops()` in the `DropletUtils` package. 
This method was then adopted, with some modifications, as the default cell filtering method used by Cell Ranger.

> Here we will use the [`emptyDropsCellRanger()` function](https://rdrr.io/github/MarioniLab/DropletUtils/man/emptyDropsCellRanger.html) to perform filtering that more closely matches the Cell Ranger implementation.

Reviewer: I was wondering about whether the `bioc` rdrr link should be used ( https://rdrr.io/bioc/DropletUtils/) instead but it looks like that site is very out of date (several minor versions behind) and doesn't include this function!
I'll note though that this old version has the best maintainer contact info of all versions. It's not close.

We can look at just the rows without `NA` values by selected the ones where the FDR (which we will use again soon), is not `NA`.

```{r droplet stats}
# view rows where FDR is not `NA`
droplet_df[!is.na(droplet_df$FDR), ]
```
You will notice that some cells with high counts also have `NA` values for many statistics.
In those cases, `NA` values are actually present _because_ of the high counts - `emptyDropsCellRanger()` automatically assumed cells were present, so they were also not tested. 

Now we can filter our `raw_sce` object _by column_ to only keep the cells with a small FDR: those that are quite unlikely to be empty droplets. 

```{r filter emptydrops, live=TRUE}

> # filter droplets using `which` to prevent NA trouble

Reviewer: I want to beef this up a tiny bit, although now I'm writing "`which` which", and we're back to https://github.com/AlexsLemonade/training-modules/pull/586#discussion_r1043887822! - 
But I do want to maybe get this idea across more explicitly?

```suggestion
# filter droplets using `which` which will ignore NAs
```

## Additional Quality control

> In addition to filtering out empty droplets, we also will want to filter out "real" cells that may have been damaged during library preparation.

Reviewer: Can we nail down "real" (with quotes) a bit more precisely?

## Additional Quality control

> In addition to filtering out empty droplets, we also will want to filter out cells that may have been damaged during library preparation.

Reviewer 2: Part of me thinks it might be helpful to state here that you also want to consider number of genes detected in a cell. I don't want them to get the impression that mitochondrial content is the only thing that should be considered. Even though you talk about how they can be correlated and how miQC filters both, I think this intro should address that both a low number of genes detected and high mito content indicate a compromised cell.

Author: Thanks for this suggestion, I added some clauses here to note that both of these features are important, and I think it motivates the model better!

These will often be characterized by a high proportion of mitochondrial transcripts.
When a cell ruptures, cytoplasmic transcripts will leak out, but mitochondrial transcripts, still protected by the mitochondrial membrane, may remain.
As a consequence, there will be an over-abundance of mitochondrial reads.

Our first step then, is create a vector of the mitochondrial genes that are present in our dataset.
The mitochondrial file we defined during setup (`mito_file`) is a TSV file containing all of the human mitochondrial genes with additional annotation information for each gene, such as the gene location and alternative names. 

> (For more detail on the steps we took to create this file, you can look at [one of our setup notebooks](https://github.com/AlexsLemonade/training-modules/blob/master/scRNA-seq-advanced/setup/mito_gene_lists.Rmd))

Reviewer: Misc comment for whether this should be `master` or at a specific commit just in case?

Author: Ideally it is tagged to this release, but that is circular, so `master` should work for now. I don't expect this to move, but who knows. (Worth noting that we remove the setup notebooks when building user directories, so my initial idea of telling people to just look within their folders won't work.)

The `mito` name is important in that it is the name that will be expected by a later function.
(We could define more subsets, but for now this one will do.)

```{r per cell QC, live=TRUE}
filtered_sce <- scuttle::addPerCellQC(filtered_sce,
                                      subsets = list(mito = mito_genes))
```

Now we can look at the colData to see what was added:

> ```{r view colData stats}

Reviewer: live?

The `mito` name is important in that it is the name that will be expected by a later function.
(We could define more subsets, but for now this one will do.)

```{r per cell QC, live=TRUE}
filtered_sce <- scuttle::addPerCellQC(filtered_sce,
                                      subsets = list(mito = mito_genes))
```

Now we can look at the colData to see what was added:

```{r view colData stats}
head(colData(filtered_sce))
```

> 

Reviewer 2: There is a typo on line 273 that I can't comment on `plots` should be `plot` towards the end of the line.

**Always look at your data!**

```{r miQC plotModel, live=TRUE}
# plot the miQC model
miQC::plotModel(filtered_sce, miqc_model) + 
  theme_bw()
```

We can now filter our data based on the probability compromised as calculated from the model.
But before we do that, we might want to quickly plot to see what would be filtered out with a given cutoff, using the `plotFiltering()` function.
The default is to exclude cells that have a posterior probability of 0.75 or greater of being compromised.
We stick with that default, but for clarity, we will also include it in our code!

```{r miQC plotFiltering}
# look at miQC filtering
miQC::plotFiltering(filtered_sce, miqc_model, posterior_cutoff = 0.75) + 
  theme_bw()
```

In this case, the line between the cells to be kept and those that will be removed seems to correspond to a mitochondrial percentage of about 12.5%, but note that this will not always be constant.
The cutoff point can vary for different numbers of unique genes within a sample, and it will certainly vary among samples!

At this point, we can perform the actual filtering using the `filterCells()` function, giving us a further filtered SCE object.

> ```{r miQC filtercells,live=TRUE}

Reviewer: ```suggestion
```{r miQC filtercells, live=TRUE}
```

This creates a new "assay" in the `normalized_sce` object, `logcounts`, which contains the normalized count values for each cell and gene.

> (The data here are not _actually_ the log of the counts, since we also applied the scaling factors, but that is what we have.)

Reviewer: ```suggestion
(The data here are not _actually_ the log of the counts, since we also applied the scaling factors, but that name is used for historicial reasons.)
```

Let's take a look:

> ```{r normalized sce}

Reviewer: live?

Note that for these reduced-dimensionality matrices, the rows are the cells and the columns are the PC dimensions.

> ```{r getReducedDim}

Reviewer: live?

We will now save our filtered and normalized object, including the dimension reduction and clustering results to an `RDS` file, using the file path that we defined at the start of the notebook.
If we were to want to return to this data, we could load this file directly into a new R session and not have to repeat the processing that we have done up to this point.

```{r save SCE, live=TRUE}

> readr::write_rds(normalized_sce, file = output_sce_file, compress="gz")

Reviewer: We should probably mention the compression here explicitly in text or comment
